asllib/aslspec/spec.ml

```diff
@@ -291,6 +323,323 @@ type t = {
+        rule_elements
+    in
+    (* Second, aggregate consecutive Cases elements. *)
+    List.fold_right
+      (fun rule_elem suffix_elements ->
+        match (rule_elem, suffix_elements) with
+        | Judgment _, _ | Rule.Cases _, Judgment _ :: _ ->
+            rule_elem :: suffix_elements
+        | Rule.Cases cases, [] -> [ Rule.Cases cases ]
+        | Rule.Cases cases, Cases next_cases :: tail ->
+            Rule.Cases (cases @ next_cases) :: tail)
+      rule_elements []
+
+  (** Transforms relations in three ways:
+      {ul
+       {- Aggregates consecutive [Case] blocks into a single [Case] block. }
+       {- Resolves [Unresolved] in applicators to either [Node_Relation] or
+          [Node_TypeVariant]
+       }
+       {- Transforms the conclusion judgment from an expression to a transition.
+          That is, given
+          {[
+            relation r(a: A) -> (b:B) =
+              p1; ... pk;
+              --
+              c_expr;
+          ]}
+          we transform it to
+          {[
+            relation r(a: A) -> (b:B) =
+              p1; ... pk;
+              --
+              r(a) -> c_expr;
+          ]}
+          The result is the re-written AST along with an updated map from
+          identifiers to definition nodes, updated for the newly created
+          relation nodes.
+       }
+      } *)
+  let resolve ast id_to_defining_node =
+    let open Rule in
+    let ast =
+      List.map
+        (fun elem ->
+          match elem with
+          | Elem_Type _ | Elem_Constant _ | Elem_RenderTypes _
+          | Elem_RenderRule _ ->
+              elem
+          | Elem_Relation def ->
+              let resolved_def =
+                match def.rule_opt with
+                | None -> def
+                | Some elements ->
+                    let elements = aggregate_case_blocks elements in
+                    let conclusion_lhs = lhs_of_conclusion def in
+                    let resolved_elements =
+                      List.map
+                        (fun rule_elem ->
+                          resolve_rule_element id_to_defining_node
+                            conclusion_lhs rule_elem)
+                        elements
+                    in
+                    { def with rule_opt = Some resolved_elements }
+              in
+              Elem_Relation resolved_def)
+        ast
+    in
+    (* Since we generated new relation nodes, we need to re-generate id_to_defining_node. *)
+    let definition_nodes = list_definition_nodes ast in
+    let id_to_defining_node = make_id_to_definition_node definition_nodes in
+    (ast, id_to_defining_node)
+end
+
+(** A module for expanding rules with cases into multiple rules without cases.
+*)
+module ExpandRules = struct
+  open Rule
+
+  type expanded_rule = {
+    name_opt : string option;
+        (** The optional name of the expanded rule, if it originated from a
+            case. *)
+    category_opt : Relation.relation_category option;
```

Comment:
I'm a bit confused by `category_opt` - in the code below, it seems to be effectively "constant". Does it need to be in this type, or could it be factored out?